docs/proposals/vip-3.md

```diff
@@ -0,0 +1,226 @@
+
+Because it is supported in Rust, and that should make the integration easier. An alternative would be to have the Java 
+server communicate with the Rust server via some other mechanism (loopback socket, pipes, etc) but then that would mean
+adding lots of forks in the code, and might also be less efficient. Using a non-JNI mechanism would also mean that just
+enabling the Rust server (even if it's not used yet) would already alter the behavior of the main code, thus creating
+more risk. In order to work in an incremental manner, where the Rust code can be developed in small chunks and hidden
+behind a config flag until ready, it is preferable to keep this interface the same.
+
+In the end state, JNI would only be used by the Java code responsible for the ingestion, and the read path would be
+fully JNI-free. Also, it should be noted that the Rust server will not be a separate child process spawned by the Java
+process, but rather fully embedded in the Java process. The Rust server will be activated by a JNI call initiated by the
+Java server, which spins up the gRPC Handler on the Rust side, along with any port and threads it needs to function.
+
+Doing JNI properly requires packaging a jar with code compiled for multiple architectures (at least Mac and Linux for
+the time being). There is some complexity there, but it can be done.
+
+### Quota
+
+The read quota will need to be re-implemented in Rust. There are essentially two main approaches we can consider:
+
+1. Let the Rust server and Java server share a quota amount between them, so that no matter if the traffic comes in via
+   the Java port or the Rust port, it will be accounted for and the configured limit will be respected. This requires
+   having some kind of synchronization between the two sides (either at the granularity of each request, which is likely
+   too expensive, or by reserving some allotment periodically, which is complicated).
+2. Let the Rust server and Java server each have the full quota amount and not need any interaction between each other 
+   for sharing it. This is much simpler to implement, but the downside is that if clients were to perfectly balance
+   their traffic across the two ports, they could get up to twice their configured quota.
+
+Since the migration is expected to be controlled by the operator, and should only be activated for users of the Fast
+Client (i.e. not for users that run a mix of Thin Clients and Fast Clients), the edge case solved by the more 
+complicated solution 1 does not seem likely to come up. Therefore, the proposal is to pick the simpler solution 2.
+
+Also, it can be considered whether quota should be part of the MVP at all. If the Rust server mode will be tried in 
+dedicated clusters at first, then quota may not even be a must-have (though of course it will be needed eventually).
+
+### Cluster Discovery
+
+The general idea is that when the Rust mode is activated in a server, that server would announce that it has a gRPC 
```

Comment:
For metrics (especially read metrics), now it is inside Rust Server then? (Since for the FC->Rust path no Java is touched) 
Just confirming are we going to do metrics aggregation just like what II is doing today?

Reply:
I am thinking to do it similarly to II, yes. If you have feedback on that approach (either for or against), please share!

Reply:
I think it depends on if LI sidecar container support is fully there or not, for now it is not there, so we have to consolidate metrics into single Java process for our InGraph to consume. 
For OSS Venice, either way is fine, we can expose endpoints to let external consumer to consume metrics from Rust Process.

Reply:
I added a metrics section to the doc.